src/java/org/apache/cassandra/db/monitoring/MonitoringTask.java

```diff
@@ -328,31 +379,126 @@ protected abstract static class Operation
          * this is set lazily as it takes time to build the query CQL */
         private String name;
 
+        /**
+         * creation time of this Operation object, in ms,
+         * this is different from operation's creationTimeNanos
+         * which does not follow wall clock and is useless for
+         * reporting purposes e.g. in virtual tables
+         */
+        private final long timestamp;
+
+        // optional keyspace and table this operation acts on
+        // used upon deserialisation
+        private String keyspace;
+        private String table;
+        private boolean crossNode;
+
         Operation(Monitorable operation, long failedAtNanos)
         {
             this.operation = operation;
             numTimesReported = 1;
             totalTimeNanos = failedAtNanos - operation.creationTimeNanos();
             minTime = totalTimeNanos;
             maxTime = totalTimeNanos;
+            timestamp = Clock.Global.currentTimeMillis();
         }
 
+        void add(Operation operation)
+        {
+            numTimesReported++;
+            totalTimeNanos += operation.totalTimeNanos;
+            maxTime = Math.max(maxTime, operation.maxTime);
+            minTime = Math.min(minTime, operation.minTime);
+        }
+
+        public abstract String getLogMessage();
+
+        @JsonProperty
         public String name()
         {
             if (name == null)
                 name = operation.name();
             return name;
         }
 
-        void add(Operation operation)
+        @JsonProperty
+        public String keyspace()
         {
-            numTimesReported++;
-            totalTimeNanos += operation.totalTimeNanos;
-            maxTime = Math.max(maxTime, operation.maxTime);
-            minTime = Math.min(minTime, operation.minTime);
+            if (operation != null)
+            {
+                String monitored = operation.monitoredOnKeyspace();
+                if (monitored != null)
+                    return monitored;
+            }
+            return keyspace;
         }
 
-        public abstract String getLogMessage();
+        @JsonProperty
+        public String table()
+        {
+            if (operation != null)
+            {
+                String monitored = operation.monitoredOnTable();
+                if (monitored != null)
+                    return monitored;
+            }
+            return table;
+        }
+
+        @JsonProperty
+        public boolean isCrossNode()
+        {
+            if (operation != null)
+                return operation.isCrossNode();
+
+            return crossNode;
+        }
+
+        @JsonProperty
+        public int numTimesReported()
+        {
+            return numTimesReported;
+        }
+
+        @JsonProperty
+        public long totalTimeNanos()
+        {
+            return totalTimeNanos;
+        }
+
+        @JsonProperty
+        public long maxTime()
```

Comment:
can we add time unit to the variable name?

Reply:
sure